File: backend/routes/analytics.py
from fastapi import APIRouter, HTTPException, status, Depends, Query
from typing import List, Optional, Dict, Any
from datetime import datetime, timedelta, date
from ..models.user import User
from ..auth.dependencies import get_current_user, require_admin_or_manager
from ..database.mongodb import DatabaseOperations
import logging

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/analytics", tags=["analytics"])

@router.get("/dashboard")
async def get_dashboard_analytics(
    current_user: User = Depends(get_current_user)
):
    """Get dashboard analytics data"""
    try:
        # Time range for analysis
        end_date = datetime.utcnow()
        start_date = end_date - timedelta(days=30)
        
        # User's time tracking stats
        user_pipeline = [
            {
                "$match": {
                    "user_id": current_user.id,
                    "start_time": {"$gte": start_date, "$lte": end_date}
                }
            },
            {
                "$group": {
                    "_id": None,
                    "total_hours": {"$sum": "$duration"},
                    "total_entries": {"$sum": 1},
                    "avg_session": {"$avg": "$duration"},
                    "projects": {"$addToSet": "$project_id"}
                }
            }
        ]
        
        user_stats = await DatabaseOperations.aggregate("time_entries", user_pipeline)
        user_data = user_stats[0] if user_stats else {
            "total_hours": 0,
            "total_entries": 0,
            "avg_session": 0,
            "projects": []
        }
        
        # Convert seconds to hours
        user_data["total_hours"] = user_data["total_hours"] / 3600
        user_data["avg_session"] = user_data["avg_session"] / 3600
        user_data["projects_count"] = len(user_data["projects"])
        
        # Daily productivity trend (last 7 days)
        daily_pipeline = [
            {
                "$match": {
                    "user_id": current_user.id,
                    "start_time": {"$gte": end_date - timedelta(days=7), "$lte": end_date}
                }
            },
            {
                "$group": {
                    "_id": {
                        "$dateToString": {
                            "format": "%Y-%m-%d",
                            "date": "$start_time"
                        }
                    },
                    "hours": {"$sum": "$duration"},
                    "activity": {"$avg": "$activity_level"}
                }
            },
            {"$sort": {"_id": 1}}
        ]
        
        daily_data = await DatabaseOperations.aggregate("time_entries", daily_pipeline)
        
        # Format daily data
        productivity_trend = []
        for day in daily_data:
            productivity_trend.append({
                "date": day["_id"],
                "hours": round(day["hours"] / 3600, 2),
                "activity": round(day["activity"] or 0, 1)
            })
        
        # Project breakdown
        project_pipeline = [
            {
                "$match": {
                    "user_id": current_user.id,
                    "start_time": {"$gte": start_date, "$lte": end_date}
                }
            },
            {
                "$group": {
                    "_id": "$project_id",
                    "hours": {"$sum": "$duration"},
                    "entries": {"$sum": 1},
                    "avg_activity": {"$avg": "$activity_level"}
                }
            },
            {"$sort": {"hours": -1}},
            {"$limit": 10}
        ]
        
        project_data = await DatabaseOperations.aggregate("time_entries", project_pipeline)
        
        # Get project names and format data
        project_breakdown = []
        for project in project_data:
            project_info = await DatabaseOperations.get_document("projects", {"id": project["_id"]})
            project_breakdown.append({
                "project_id": project["_id"],
                "project_name": project_info["name"] if project_info else "Unknown",
                "hours": round(project["hours"] / 3600, 2),
                "entries": project["entries"],
                "avg_activity": round(project["avg_activity"] or 0, 1)
            })
        
        return {
            "user_stats": user_data,
            "productivity_trend": productivity_trend,
            "project_breakdown": project_breakdown,
            "period": {
                "start_date": start_date.isoformat(),
                "end_date": end_date.isoformat()
            }
        }
        
    except Exception as e:
        logger.error(f"Dashboard analytics error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to get dashboard analytics"
        )

@router.get("/team")
async def get_team_analytics(
    start_date: Optional[date] = None,
    end_date: Optional[date] = None,
    current_user: User = Depends(require_admin_or_manager)
):
    """Get team analytics (admin/manager only)"""
    try:
        if not start_date:
            start_date = (datetime.utcnow() - timedelta(days=30)).date()
        if not end_date:
            end_date = datetime.utcnow().date()
        
        start_datetime = datetime.combine(start_date, datetime.min.time())
        end_datetime = datetime.combine(end_date, datetime.max.time())
        
        # Team productivity stats
        team_pipeline = [
            {
                "$match": {
                    "start_time": {"$gte": start_datetime, "$lte": end_datetime}
                }
            },
            {
                "$group": {
                    "_id": "$user_id",
                    "total_hours": {"$sum": "$duration"},
                    "total_entries": {"$sum": 1},
                    "avg_activity": {"$avg": "$activity_level"},
                    "projects": {"$addToSet": "$project_id"}
                }
            }
        ]
        
        team_data = await DatabaseOperations.aggregate("time_entries", team_pipeline)
        
        # Get user details and format data
        team_stats = []
        for member in team_data:
            user_info = await DatabaseOperations.get_document("users", {"id": member["_id"]})
            if user_info:
                team_stats.append({
                    "user_id": member["_id"],
                    "user_name": user_info["name"],
                    "user_role": user_info["role"],
                    "total_hours": round(member["total_hours"] / 3600, 2),
                    "total_entries": member["total_entries"],
                    "avg_activity": round(member["avg_activity"] or 0, 1),
                    "projects_count": len(member["projects"])
                })
        
        # Sort by total hours
        team_stats.sort(key=lambda x: x["total_hours"], reverse=True)
        
        # Daily team productivity
        daily_team_pipeline = [
            {
                "$match": {
                    "start_time": {"$gte": start_datetime, "$lte": end_datetime}
                }
            },
            {
                "$group": {
                    "_id": {
                        "$dateToString": {
                            "format": "%Y-%m-%d",
                            "date": "$start_time"
                        }
                    },
                    "total_hours": {"$sum": "$duration"},
                    "avg_activity": {"$avg": "$activity_level"},
                    "active_users": {"$addToSet": "$user_id"}
                }
            },
            {"$sort": {"_id": 1}}
        ]
        
        daily_team_data = await DatabaseOperations.aggregate("time_entries", daily_team_pipeline)
        
        daily_productivity = []
        for day in daily_team_data:
            daily_productivity.append({
                "date": day["_id"],
                "total_hours": round(day["total_hours"] / 3600, 2),
                "avg_activity": round(day["avg_activity"] or 0, 1),
                "active_users": len(day["active_users"])
            })
        
        # Project analytics
        project_analytics_pipeline = [
            {
                "$match": {
                    "start_time": {"$gte": start_datetime, "$lte": end_datetime}
                }
            },
            {
                "$group": {
                    "_id": "$project_id",
                    "total_hours": {"$sum": "$duration"},
                    "team_members": {"$addToSet": "$user_id"},
                    "avg_activity": {"$avg": "$activity_level"}
                }
            },
            {"$sort": {"total_hours": -1}}
        ]
        
        project_analytics = await DatabaseOperations.aggregate("time_entries", project_analytics_pipeline)
        
        # Get project details
        project_stats = []
        for project in project_analytics:
            project_info = await DatabaseOperations.get_document("projects", {"id": project["_id"]})
            if project_info:
                project_stats.append({
                    "project_id": project["_id"],
                    "project_name": project_info["name"],
                    "total_hours": round(project["total_hours"] / 3600, 2),
                    "team_members": len(project["team_members"]),
                    "avg_activity": round(project["avg_activity"] or 0, 1),
                    "budget": project_info.get("budget", 0),
                    "spent": project_info.get("spent", 0)
                })
        
        return {
            "team_stats": team_stats,
            "daily_productivity": daily_productivity,
            "project_stats": project_stats,
            "summary": {
                "total_team_hours": sum(member["total_hours"] for member in team_stats),
                "avg_team_activity": sum(member["avg_activity"] for member in team_stats) / len(team_stats) if team_stats else 0,
                "active_projects": len(project_stats),
                "team_size": len(team_stats)
            },
            "period": {
                "start_date": start_date.isoformat(),
                "end_date": end_date.isoformat()
            }
        }
        
    except Exception as e:
        logger.error(f"Team analytics error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to get team analytics"
        )

@router.get("/productivity")
async def get_productivity_analytics(
    period: str = Query("week", enum=["day", "week", "month"]),
    current_user: User = Depends(get_current_user)
):
    """Get productivity analytics with different time periods"""
    try:
        # Calculate date range based on period
        end_date = datetime.utcnow()
        if period == "day":
            start_date = end_date - timedelta(days=1)
            group_format = "%Y-%m-%d %H:00"
        elif period == "week":
            start_date = end_date - timedelta(days=7)
            group_format = "%Y-%m-%d"
        else:  # month
            start_date = end_date - timedelta(days=30)
            group_format = "%Y-%m-%d"
        
        # Productivity trend analysis
        productivity_pipeline = [
            {
                "$match": {
                    "user_id": current_user.id,
                    "start_time": {"$gte": start_date, "$lte": end_date}
                }
            },
            {
                "$group": {
                    "_id": {
                        "$dateToString": {
                            "format": group_format,
                            "date": "$start_time"
                        }
                    },
                    "hours": {"$sum": "$duration"},
                    "activity": {"$avg": "$activity_level"},
                    "entries": {"$sum": 1},
                    "mouse_clicks": {"$sum": {"$ifNull": ["$mouse_clicks", 0]}},
                    "keyboard_strokes": {"$sum": {"$ifNull": ["$keyboard_strokes", 0]}}
                }
            },
            {"$sort": {"_id": 1}}
        ]
        
        productivity_data = await DatabaseOperations.aggregate("time_entries", productivity_pipeline)
        
        # Format productivity data
        productivity_chart = []
        for data_point in productivity_data:
            productivity_chart.append({
                "timestamp": data_point["_id"],
                "hours": round(data_point["hours"] / 3600, 2),
                "activity_level": round(data_point["activity"] or 0, 1),
                "entries": data_point["entries"],
                "mouse_clicks": data_point["mouse_clicks"],
                "keyboard_strokes": data_point["keyboard_strokes"]
            })
        
        # Calculate productivity score
        total_hours = sum(point["hours"] for point in productivity_chart)
        avg_activity = sum(point["activity_level"] for point in productivity_chart) / len(productivity_chart) if productivity_chart else 0
        
        productivity_score = min(100, (avg_activity * 0.7 + (total_hours / (len(productivity_chart) * 8)) * 100 * 0.3))
        
        return {
            "productivity_chart": productivity_chart,
            "productivity_score": round(productivity_score, 1),
            "total_hours": round(total_hours, 2),
            "avg_activity": round(avg_activity, 1),
            "period": period,
            "date_range": {
                "start": start_date.isoformat(),
                "end": end_date.isoformat()
            }
        }
        
    except Exception as e:
        logger.error(f"Productivity analytics error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to get productivity analytics"
        )

@router.get("/reports/custom")
async def generate_custom_report(
    start_date: date,
    end_date: date,
    user_ids: Optional[List[str]] = Query(None),
    project_ids: Optional[List[str]] = Query(None),
    current_user: User = Depends(require_admin_or_manager)
):
    """Generate custom analytics report"""
    try:
        start_datetime = datetime.combine(start_date, datetime.min.time())
        end_datetime = datetime.combine(end_date, datetime.max.time())
        
        # Build query
        match_query = {
            "start_time": {"$gte": start_datetime, "$lte": end_datetime}
        }
        
        if user_ids:
            match_query["user_id"] = {"$in": user_ids}
        if project_ids:
            match_query["project_id"] = {"$in": project_ids}
        
        # Comprehensive analytics pipeline
        analytics_pipeline = [
            {"$match": match_query},
            {
                "$group": {
                    "_id": {
                        "user_id": "$user_id",
                        "project_id": "$project_id",
                        "date": {
                            "$dateToString": {
                                "format": "%Y-%m-%d",
                                "date": "$start_time"
                            }
                        }
                    },
                    "hours": {"$sum": "$duration"},
                    "activity": {"$avg": "$activity_level"},
                    "entries": {"$sum": 1}
                }
            }
        ]
        
        detailed_data = await DatabaseOperations.aggregate("time_entries", analytics_pipeline)
        
        # Process and format the data
        report_data = []
        for entry in detailed_data:
            user_info = await DatabaseOperations.get_document("users", {"id": entry["_id"]["user_id"]})
            project_info = await DatabaseOperations.get_document("projects", {"id": entry["_id"]["project_id"]})
            
            report_data.append({
                "date": entry["_id"]["date"],
                "user_name": user_info["name"] if user_info else "Unknown",
                "project_name": project_info["name"] if project_info else "Unknown",
                "hours": round(entry["hours"] / 3600, 2),
                "activity_level": round(entry["activity"] or 0, 1),
                "entries": entry["entries"]
            })
        
        # Calculate summary statistics
        total_hours = sum(entry["hours"] for entry in report_data)
        avg_activity = sum(entry["activity_level"] for entry in report_data) / len(report_data) if report_data else 0
        
        return {
            "report_data": report_data,
            "summary": {
                "total_hours": round(total_hours, 2),
                "avg_activity": round(avg_activity, 1),
                "total_entries": sum(entry["entries"] for entry in report_data),
                "date_range": f"{start_date} to {end_date}"
            },
            "generated_at": datetime.utcnow().isoformat()
        }
        
    except Exception as e:
        logger.error(f"Custom report error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to generate custom report"
        )

File: backend/routes/analytics.js
const express = require("express");
const TimeEntry = require("../models/TimeEntry");
const Project = require("../models/Project");
const Task = require("../models/Task");
const User = require("../models/User");
const { authMiddleware, requireManager } = require("../middleware/auth");

const router = express.Router();

// Get dashboard analytics - FIXED: Organization-scoped
router.get("/dashboard", authMiddleware, async (req, res) => {
  try {
    const { period = "week" } = req.query;

    // Calculate date range based on period
    const endDate = new Date();
    let startDate = new Date();

    switch (period) {
      case "day":
        startDate.setDate(endDate.getDate() - 1);
        break;
      case "week":
        startDate.setDate(endDate.getDate() - 7);
        break;
      case "month":
        startDate.setMonth(endDate.getMonth() - 1);
        break;
      case "year":
        startDate.setFullYear(endDate.getFullYear() - 1);
        break;
      default:
        startDate.setDate(endDate.getDate() - 7);
    }

    // FIXED: Get time entries for the period filtered by organization
    const entries = await TimeEntry.find({
      user_id: req.user.id,
      organizationId: req.user.organizationId, // CRITICAL FIX
      start_time: { $gte: startDate, $lte: endDate },
      end_time: { $ne: null },
    });

    // Calculate user stats
    const totalHours =
      entries.reduce((sum, entry) => sum + entry.duration, 0) / 3600;
    const totalEntries = entries.length;
    const avgActivity =
      entries.length > 0
        ? entries.reduce((sum, entry) => sum + entry.activity_level, 0) /
          entries.length
        : 0;

    // Project breakdown
    const projectBreakdown = entries.reduce((acc, entry) => {
      if (!acc[entry.project_id]) {
        acc[entry.project_id] = {
          project_id: entry.project_id,
          hours: 0,
          entries: 0,
        };
      }
      acc[entry.project_id].hours += entry.duration / 3600;
      acc[entry.project_id].entries += 1;
      return acc;
    }, {});

    // Productivity trend (daily aggregation)
    const dailyTrend = {};
    entries.forEach((entry) => {
      const date = entry.start_time.toISOString().split("T")[0];
      if (!dailyTrend[date]) {
        dailyTrend[date] = {
          date,
          hours: 0,
          activity: 0,
          entries: 0,
        };
      }
      dailyTrend[date].hours += entry.duration / 3600;
      dailyTrend[date].activity += entry.activity_level;
      dailyTrend[date].entries += 1;
    });

    // Calculate average activity per day
    Object.values(dailyTrend).forEach((day) => {
      if (day.entries > 0) {
        day.activity = day.activity / day.entries;
      }
    });

    console.log(
      `Dashboard analytics for user ${req.user.email} in organization ${req.user.organizationId}`
    );

    res.json({
      user_stats: {
        total_hours: Math.round(totalHours * 100) / 100,
        total_entries: totalEntries,
        avg_activity: Math.round(avgActivity * 100) / 100,
        productive_days: Object.keys(dailyTrend).length,
      },
      productivity_trend: Object.values(dailyTrend).sort(
        (a, b) => new Date(a.date) - new Date(b.date)
      ),
      project_breakdown: Object.values(projectBreakdown),
      period,
    });
  } catch (error) {
    console.error("Get dashboard analytics error:", error);
    res.status(500).json({ error: "Internal server error" });
  }
});

// Get team analytics (manager/admin only) - FIXED: Organization-scoped
router.get("/team", authMiddleware, requireManager, async (req, res) => {
  try {
    const { period = "week" } = req.query;

    // Calculate date range
    const endDate = new Date();
    let startDate = new Date();

    switch (period) {
      case "day":
        startDate.setDate(endDate.getDate() - 1);
        break;
      case "week":
        startDate.setDate(endDate.getDate() - 7);
        break;
      case "month":
        startDate.setMonth(endDate.getMonth() - 1);
        break;
      case "year":
        startDate.setFullYear(endDate.getFullYear() - 1);
        break;
      default:
        startDate.setDate(endDate.getDate() - 7);
    }

    // FIXED: Get all time entries for the period filtered by organization
    const entries = await TimeEntry.find({
      organizationId: req.user.organizationId, // CRITICAL FIX
      start_time: { $gte: startDate, $lte: endDate },
      end_time: { $ne: null },
    });

    // Team statistics
    const teamStats = entries.reduce((acc, entry) => {
      if (!acc[entry.user_id]) {
        acc[entry.user_id] = {
          user_id: entry.user_id,
          total_hours: 0,
          total_entries: 0,
          avg_activity: 0,
          projects: new Set(),
        };
      }

      acc[entry.user_id].total_hours += entry.duration / 3600;
      acc[entry.user_id].total_entries += 1;
      acc[entry.user_id].avg_activity += entry.activity_level;
      acc[entry.user_id].projects.add(entry.project_id);

      return acc;
    }, {});

    // Calculate averages
    Object.values(teamStats).forEach((user) => {
      user.avg_activity =
        user.total_entries > 0 ? user.avg_activity / user.total_entries : 0;
      user.projects_count = user.projects.size;
      user.projects = Array.from(user.projects);
    });

    // Daily productivity
    const dailyProductivity = {};
    entries.forEach((entry) => {
      const date = entry.start_time.toISOString().split("T")[0];
      if (!dailyProductivity[date]) {
        dailyProductivity[date] = {
          date,
          total_hours: 0,
          avg_activity: 0,
          active_users: new Set(),
        };
      }
      dailyProductivity[date].total_hours += entry.duration / 3600;
      dailyProductivity[date].avg_activity += entry.activity_level;
      dailyProductivity[date].active_users.add(entry.user_id);
    });

    // Calculate daily averages
    Object.values(dailyProductivity).forEach((day) => {
      const entriesCount = entries.filter(
        (e) => e.start_time.toISOString().split("T")[0] === day.date
      ).length;
      day.avg_activity = entriesCount > 0 ? day.avg_activity / entriesCount : 0;
      day.active_users = day.active_users.size;
    });

    // Project statistics - FIXED: Organization-scoped
    const projectStats = entries.reduce((acc, entry) => {
      if (!acc[entry.project_id]) {
        acc[entry.project_id] = {
          project_id: entry.project_id,
          total_hours: 0,
          contributors: new Set(),
        };
      }
      acc[entry.project_id].total_hours += entry.duration / 3600;
      acc[entry.project_id].contributors.add(entry.user_id);
      return acc;
    }, {});

    Object.values(projectStats).forEach((project) => {
      project.contributors = project.contributors.size;
    });

    // Summary statistics
    const summary = {
      total_hours:
        entries.reduce((sum, entry) => sum + entry.duration, 0) / 3600,
      total_entries: entries.length,
      active_users: Object.keys(teamStats).length,
      active_projects: Object.keys(projectStats).length,
      avg_activity:
        entries.length > 0
          ? entries.reduce((sum, entry) => sum + entry.activity_level, 0) /
            entries.length
          : 0,
    };

    console.log(
      `Team analytics for organization ${req.user.organizationId}: ${summary.active_users} users, ${summary.active_projects} projects`
    );

    res.json({
      team_stats: Object.values(teamStats),
      daily_productivity: Object.values(dailyProductivity).sort(
        (a, b) => new Date(a.date) - new Date(b.date)
      ),
      project_stats: Object.values(projectStats),
      summary,
      period,
    });
  } catch (error) {
    console.error("Get team analytics error:", error);
    res.status(500).json({ error: "Internal server error" });
  }
});

// Get productivity analytics - FIXED: Organization-scoped
router.get("/productivity", authMiddleware, async (req, res) => {
  try {
    const { period = "week", user_id } = req.query;

    // Calculate date range
    const endDate = new Date();
    let startDate = new Date();

    switch (period) {
      case "day":
        startDate.setDate(endDate.getDate() - 1);
        break;
      case "week":
        startDate.setDate(endDate.getDate() - 7);
        break;
      case "month":
        startDate.setMonth(endDate.getMonth() - 1);
        break;
      case "year":
        startDate.setFullYear(endDate.getFullYear() - 1);
        break;
      default:
        startDate.setDate(endDate.getDate() - 7);
    }

    // Query filter - FIXED: Always include organization filter
    const query = {
      organizationId: req.user.organizationId, // CRITICAL FIX
      start_time: { $gte: startDate, $lte: endDate },
      end_time: { $ne: null },
    };

    // If user_id is provided and user is manager/admin, filter by user
    if (user_id && ["admin", "manager"].includes(req.user.role)) {
      // FIXED: Ensure user_id belongs to same organization
      const targetUser = await User.findOne({
        id: user_id,
        organizationId: req.user.organizationId,
      });

      if (!targetUser) {
        return res
          .status(404)
          .json({ error: "User not found in your organization" });
      }

      query.user_id = user_id;
    } else {
      query.user_id = req.user.id;
    }

    const entries = await TimeEntry.find(query);

    // Productivity chart data (hourly breakdown)
    const hourlyData = {};
    entries.forEach((entry) => {
      const hour = entry.start_time.getHours();
      if (!hourlyData[hour]) {
        hourlyData[hour] = {
          hour,
          total_duration: 0,
          avg_activity: 0,
          entries_count: 0,
        };
      }
      hourlyData[hour].total_duration += entry.duration;
      hourlyData[hour].avg_activity += entry.activity_level;
      hourlyData[hour].entries_count += 1;
    });

    // Calculate averages
    Object.values(hourlyData).forEach((hour) => {
      hour.avg_activity =
        hour.entries_count > 0 ? hour.avg_activity / hour.entries_count : 0;
    });

    // Overall productivity metrics
    const totalHours =
      entries.reduce((sum, entry) => sum + entry.duration, 0) / 3600;
    const avgActivity =
      entries.length > 0
        ? entries.reduce((sum, entry) => sum + entry.activity_level, 0) /
          entries.length
        : 0;

    // Productivity score calculation (based on activity level and hours worked)
    const targetHours =
      period === "day"
        ? 8
        : period === "week"
        ? 40
        : period === "month"
        ? 160
        : 1920;
    const hoursScore = Math.min(totalHours / targetHours, 1) * 50;
    const activityScore = (avgActivity / 100) * 50;
    const productivityScore = hoursScore + activityScore;

    res.json({
      productivity_chart: Object.values(hourlyData).sort(
        (a, b) => a.hour - b.hour
      ),
      productivity_score: Math.round(productivityScore * 100) / 100,
      total_hours: Math.round(totalHours * 100) / 100,
      avg_activity: Math.round(avgActivity * 100) / 100,
      period,
    });
  } catch (error) {
    console.error("Get productivity analytics error:", error);
    res.status(500).json({ error: "Internal server error" });
  }
});

// Generate custom report - FIXED: Organization-scoped
router.post(
  "/reports/custom",
  authMiddleware,
  requireManager,
  async (req, res) => {
    try {
      const {
        start_date,
        end_date,
        user_ids = [],
        project_ids = [],
        metrics = ["hours", "activity", "projects"],
      } = req.body;

      if (!start_date || !end_date) {
        return res
          .status(400)
          .json({ error: "Start date and end date are required" });
      }

      const startDate = new Date(start_date);
      const endDate = new Date(end_date);

      // FIXED: Always filter by organization
      const query = {
        organizationId: req.user.organizationId, // CRITICAL FIX
        start_time: { $gte: startDate, $lte: endDate },
        end_time: { $ne: null },
      };

      // Filter by users - FIXED: Ensure users belong to same organization
      if (user_ids.length > 0) {
        const orgUsers = await User.find({
          id: { $in: user_ids },
          organizationId: req.user.organizationId,
        }).select("id");

        const validUserIds = orgUsers.map((u) => u.id);
        query.user_id = { $in: validUserIds };
      }

      // Filter by projects - FIXED: Ensure projects belong to same organization
      if (project_ids.length > 0) {
        const orgProjects = await Project.find({
          id: { $in: project_ids },
          organizationId: req.user.organizationId,
        }).select("id");

        const validProjectIds = orgProjects.map((p) => p.id);
        query.project_id = { $in: validProjectIds };
      }

      const entries = await TimeEntry.find(query);

      const report = {
        date_range: {
          start: start_date,
          end: end_date,
        },
        filters: {
          users: user_ids,
          projects: project_ids,
          organization: req.user.organizationId,
        },
        metrics: {},
      };

      // Calculate requested metrics
      if (metrics.includes("hours")) {
        report.metrics.total_hours =
          entries.reduce((sum, entry) => sum + entry.duration, 0) / 3600;
        report.metrics.avg_daily_hours =
          report.metrics.total_hours /
          Math.ceil((endDate - startDate) / (1000 * 60 * 60 * 24));
      }

      if (metrics.includes("activity")) {
        report.metrics.avg_activity =
          entries.length > 0
            ? entries.reduce((sum, entry) => sum + entry.activity_level, 0) /
              entries.length
            : 0;
      }

      if (metrics.includes("projects")) {
        const projectsWorked = new Set(
          entries.map((entry) => entry.project_id)
        );
        report.metrics.projects_worked = projectsWorked.size;
        report.metrics.project_distribution = {};

        entries.forEach((entry) => {
          if (!report.metrics.project_distribution[entry.project_id]) {
            report.metrics.project_distribution[entry.project_id] = 0;
          }
          report.metrics.project_distribution[entry.project_id] +=
            entry.duration / 3600;
        });
      }

      console.log(
        `Custom report generated for organization ${req.user.organizationId}`
      );

      res.json(report);
    } catch (error) {
      console.error("Generate custom report error:", error);
      res.status(500).json({ error: "Internal server error" });
    }
  }
);

module.exports = router;
